### PR TITLE
Support polymorphism on functions

### PR DESCRIPTION
**Pull request by ThomasDutartre**
### Summary

This PR adds support for polymorphic functions (functions with the same name but different return types) in the `@Spyable` macro.

Previously, `Spyable` generated duplicate spy variable names when multiple functions shared the same name, because return types were not considered. This caused conflicts or missing spies.

### What's new

- When multiple functions share the same name but differ by return type, the return type is now included in the generated spy variable name (e.g., `doSomethingStringOptionalReturnCalls`).
- This logic only applies when name conflicts are detected — for functions with unique names, behavior remains unchanged.
- A new unit test (`VariablePrefixTestProtocol`) has been added to cover and validate all supported polymorphic cases.

### Supported return types

The following are now correctly handled in function differentiation:
- Arrays (e.g., `[String]`)
- Sets (e.g., `Set<Int?>`)
- Optionals
- Tuples (e.g., `(Int, Int)`)
- Dictionaries (e.g., `[String: Int]`)
- Simple and complex types

This ensures that spy generation is now collision-free and stable for protocols using function overloading via return types.

### CODE REVIEW

**Comment by ThomasDutartre** [comment]
https://github.com/Matejkob/swift-spyable/issues/135

File: Sources/SpyableMacro/Factories/SpyFactory.swift [hunk 1]
```diff
@@ -133,7 +137,8 @@ struct SpyFactory {
         }
 
         for functionDeclaration in functionDeclarations {
-          let variablePrefix = variablePrefixFactory.text(for: functionDeclaration)
+          let shouldAddReturnType = variablePrefixes.count(where: { $0 == (variablePrefixFactory.text(for: functionDeclaration))}) > 1
```
Comment (Matejkob) [inline Sources/SpyableMacro/Factories/SpyFactory.swift:140]: The CI is failing for Swift < 6.0 with this error:

```
cannot call value of non-function type 'Int'
```

it is due to the usage of the `count(where:)` method, which **does not exist** in Swift 5.10 (or lower). It was added to Swfit 6.0.

### Explanation

In **Swift 6**, [`count(where:)`](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0220-count-where.md) was added as a shorthand to count how many elements in a sequence match a predicate. But this method is **not available** in Swift 5.10 or earlier.

So, in Swift 5.10, you need to use the more verbose form:

```swift
let shouldAddReturnType = variablePrefixes.filter {
    $0 == variablePrefixFactory.text(for: functionDeclaration)
}.count > 1
```

This is functionally equivalent and works in both Swift 5.10 and Swift 6.

### TL;DR

Replace:

```swift
variablePrefixes.count(where: { ... })
```

with:

```swift
variablePrefixes.filter { ... }.count
```

for Swift 5.10 compatibility.

**Comment by Matejkob** [comment]
## Code Review for PR #141: Support polymorphism on functions

### Summary
This PR implements adding return type suffixes only when naming conflicts are detected. This is a good approach that maintains backward compatibility.

### ✅ Strengths

1. **Backward Compatible**: Only adds suffixes when conflicts exist, preserving existing test code
2. **Clean Implementation**: The logic is well-contained in `VariablePrefixFactory` 
3. **Comprehensive Test Coverage**: The test case covers many edge cases including arrays, sets, optionals, tuples, and dictionaries
4. **Follows Existing Pattern**: Similar to how AutoMockable.stencil handles this issue

### 🔍 Areas for Improvement

1. **Performance Consideration**
   - The conflict detection in `SpyFactory.swift` counts occurrences for every function: `variablePrefixes.count(where: { $0 == (variablePrefixFactory.text(for: functionDeclaration))}) > 1`
   - This is O(n²) complexity. Consider using a dictionary to track counts once.

2. **Return Type Formatting**
   - The string manipulation in `VariablePrefixFactory` could be more robust:
     ```swift
     returnTypeText = returnTypeText
       .replacingOccurrences(of: "?", with: "Optional")
       .removeCharacters(in: [":", "[", "]", "<", ">", "(", ")", ",", " "])
     ```
   - This might produce unexpected results for complex types like `Result<String, Error>`

3. **Missing Edge Cases**
   - What happens with generic return types? (e.g., `func foo<T>() -> T`)
   - Protocol compositions as return types (e.g., `func bar() -> Protocol1 & Protocol2`)
   - Closures as return types (e.g., `func baz() -> (Int) -> String`)

4. **Code Organization**
   - The `removeCharacters` extension is marked as `fileprivate` but could be useful elsewhere
   - Consider moving string manipulation helpers to a dedicated file

### 🐛 Potential Issues

1. **Type Name Collisions**: After character removal, different types might produce the same suffix:
   - `Set<Int>` → `SetInt`
   - `[Set: Int]` → `SetInt` (if this were valid)

2. **Optional Handling**: The special handling for optionals at the end might conflict with the earlier replacement:
   ```swift
   if returnTypeText.isOptional {
       returnTypeText.removeLast()
       returnTypeText = "Optional\(returnTypeText)"
   }
   // Later...
   .replacingOccurrences(of: "?", with: "Optional")
   ```

### 💡 Suggestions

1. Consider using a more sophisticated type name formatter that preserves type information better
2. Add tests for error-prone cases like generic types and protocol compositions
3. Consider caching the conflict detection results to avoid repeated calculations
4. Document the naming convention in the README for users

### Verdict
This is a solid implementation that solves the immediate problem. The approach is correct and the tests demonstrate it works well for common cases. The performance and edge case concerns are minor and could be addressed in future iterations.

**Recommendation**: ✅ Approve with suggestions for future improvements

**Review by Matejkob — CHANGES_REQUESTED** [review]
1. **Performance Consideration**
   - The conflict detection in `SpyFactory.swift` counts occurrences for every function: `variablePrefixes.count(where: { $0 == (variablePrefixFactory.text(for: functionDeclaration))}) > 1`
   - This is O(n²) complexity. Consider using a dictionary to track counts once.

2. **Return Type Formatting**
   - The string manipulation in `VariablePrefixFactory` could be more robust:
     ```swift
     returnTypeText = returnTypeText
       .replacingOccurrences(of: "?", with: "Optional")
       .removeCharacters(in: [":", "[", "]", "<", ">", "(", ")", ",", " "])
     ```
   - This might produce unexpected results for complex types like `Result<String, Error>`

3. **Missing Edge Cases**
   - What happens with generic return types? (e.g., `func foo<T>() -> T`)
   - Protocol compositions as return types (e.g., `func bar() -> Protocol1 & Protocol2`)
   - Closures as return types (e.g., `func baz() -> (Int) -> String`)

### 🐛 Potential Issues

1. **Type Name Collisions**: After character removal, different types might produce the same suffix:
   - `Set<Int>` → `SetInt`
   - `[Set: Int]` → `SetInt` (if this were valid)

2. **Optional Handling**: The special handling for optionals at the end might conflict with the earlier replacement:
   ```swift
   if returnTypeText.isOptional {
       returnTypeText.removeLast()
       returnTypeText = "Optional\(returnTypeText)"
   }
   // Later...
   .replacingOccurrences(of: "?", with: "Optional")
   ```

3. **Missing edge-case**:
```
protocol Foo {
  func foo(a: Int)
  func foo(a: String)
}
```

### 💡 Suggestions

1. Consider using a more sophisticated type name formatter that preserves type information better
2. Add tests for error-prone cases like generic types and protocol compositions
3. Consider caching the conflict detection results to avoid repeated calculations
5. Document the naming convention in the README for users

**Comment by ThomasDutartre** [comment]
> ### 🐛 Potential Issues
> 1. **Type Name Collisions**: After character removal, different types might produce the same suffix:
>    
>    * `Set<Int>` → `SetInt`
>    * `[Set: Int]` → `SetInt` (if this were valid)
> 2. **Optional Handling**: The special handling for optionals at the end might conflict with the earlier replacement:
>    ```swift
>    if returnTypeText.isOptional {
>        returnTypeText.removeLast()
>        returnTypeText = "Optional\(returnTypeText)"
>    }
>    // Later...
>    .replacingOccurrences(of: "?", with: "Optional")
>    ```
> 3. **Missing edge-case**:
> 
> ```
> protocol Foo {
>   func foo(a: Int)
>   func foo(a: String)
> }
> ```
> 
> ### 💡 Suggestions
> 1. Consider using a more sophisticated type name formatter that preserves type information better
> 2. Add tests for error-prone cases like generic types and protocol compositions
> 3. Consider caching the conflict detection results to avoid repeated calculations
> 4. Document the naming convention in the README for users


1.
Do you have an example to illustrate this case ?

2.
I'm handling optional this way to handle a specific case : 
```
  func compute() -> Set<String?>
  func compute() -> Set<String>?
```

Without my code, the two name would be `computeSetStringOptional`
With my code, we have `computeSetStringOptional` and `computeOptionalSetString`

3.
I handled the edge case with same parameters name but different types

**Comment by Matejkob** [comment]
@ThomasDutartre I have taken liberty to add some more changes. Thanks a lot for contributing, you have done amazing work! I'm publishing a new version of the library with your changes. It looks really good now, thanks for the contribution!

**Comment by ThomasDutartre** [comment]
> @ThomasDutartre I have taken liberty to add some more changes. Thanks a lot for contributing, you have done amazing work! I'm publishing a new version of the library with your changes. It looks really good now, thanks for the contribution!

Thank you for your changes, it's clearly better now
It was a pleasure to work on it

I think this PR is solving this issue too : https://github.com/Matejkob/swift-spyable/issues/19
